File: 0057 Insert Interval.py
# Solution from whissely
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        stack = []
        for i in range(len(intervals) - 1, -1, -1):
            stack.append(intervals[i])
            
        res = []
        while stack:
            cur_interval = stack.pop()
            # Left ouf of bounds
            if cur_interval[0] > newInterval[1]:
                res.append(newInterval)
                # Update newInterval to be cur_interval to ensure we always come into this condition because there is no more merging that will happen after newInterval has been appended so we keep appending the previous interval
                newInterval = cur_interval
            # Right out of bounds
            elif cur_interval[1] < newInterval[0]:
                res.append(cur_interval)
            else:
                newInterval = [min(cur_interval[0], newInterval[0]), max(cur_interval[1], newInterval[1])]
        # Append the last interval which will be the fully merged newInterval or the last cur_interval from the last iteration
        res.append(newInterval)
        return res

# Adapeted from Lenaxiao
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []
        s, e = newInterval[0], newInterval[1]
        add = False
        for itv in intervals:
            if not add:
                if itv[0] <= e and itv[1] >= s:
                    s = min(itv[0], s)
                    e = max(itv[1], e)
                elif itv[1] < s:
                    res.append(itv)
                else:
                    res.append([s, e])
                    res.append(itv)
                    add = True
            else:
                res.append(itv)
        if not add:
            res.append([s, e])
        
        return res

File: 0057 Insert Interval.java
// adapted from Lenaxiao
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        if (newInterval == null || newInterval.length == 0) return intervals;
        int s = newInterval[0];
        int e = newInterval[1];
        boolean add = false;
        List<int[]> res = new ArrayList<>();
        for (int[] itv : intervals) {
            if (!add) {
                if (itv[0] <= e && itv[1] >= s) { // overlapping
                    s = Math.min(itv[0], s);
                    e = Math.max(itv[1], e);
                } else if (itv[1] < s) {
                    res.add(itv);
                } else {
                    res.add(new int[] {s, e});
                    res.add(itv);
                    add = true; // no other overlapping
                }
            } else {
                res.add(itv);
            }
        }
        if (!add) {
            res.add(new int[] {s, e});
        }
        int[][] result = new int[res.size()][2];
        int i = 0;
        for (int[] itv : res) {
            result[i++] = itv;
        }
        return result;
    }
}
